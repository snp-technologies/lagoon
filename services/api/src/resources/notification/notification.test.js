// @flow

const Sql = require('./sql');

describe('Sql', () => {
  describe('selectAssignedProjectNotificationByName', () => {
    it('should create proper statement', () => {
      const input = { name: 'slack_foo', type: 'slack' };
      const ret = Sql.selectProjectNotificationByNotificationName(input);
      expect(ret).toMatchSnapshot();
    });
  });

  describe('selectProjectNotificationsWithoutAccess', () => {
    it('should create a proper statement', () => {
      const cred = {
        permissions: {
          projects: ['3', '4'],
        },
      };

      const nids = ['1'];
      const ret = Sql.selectProjectNotificationsWithoutAccess(cred, { nids });
      expect(ret).toMatchSnapshot();
    });
  });

  describe('selectNotificationsByTypeByProjectId', () => {
    it('if user, should create statement for notification_slack table with IN clause', () => {
      const cred = {
        role: 'user',
        permissions: { customers: [], projects: ['3'] },
      };

      const input = {
        type: 'slack',
        pid: 3,
      };

      const ret = Sql.selectNotificationsByTypeByProjectId(cred, input);
      expect(ret).toMatchSnapshot();
    });

    it('if admin, should create statement for notification_slack table without IN clause', () => {
      const cred = {
        role: 'admin',
        permissions: {},
      };

      const input = {
        type: 'slack',
        pid: 3,
      };

      const ret = Sql.selectNotificationsByTypeByProjectId(cred, input);
      expect(ret).toMatchSnapshot();
    });
  });

  describe('selectUnassignedNotificationsByType', () => {
    it('should create a proper query', () => {
      const ret = Sql.selectUnassignedNotificationsByType('rocketchat');
      expect(ret).toMatchSnapshot();
    });
  });

  describe('updateNotificationSlack', () => {
    it('should create a proper query', () => {
      const input = {
        name: 'n1',
        patch: {
          name: 'test',
          webhook: 'new-webhook',
        },
      };

      const ret = Sql.updateNotificationSlack(input);
      expect(ret).toMatchSnapshot();
    });
  });

  describe('selectCustomer', () => {
    it('should create a proper query', () => {
      const ret = Sql.selectNotificationSlackByName('n1');
      expect(ret).toMatchSnapshot();
    });
  });

  describe('deleteProjectNotification', () => {
    it('if not admin, should insert IN clause for project', () => {
      const cred = {
        role: 'user',
        permissions: {
          customers: [],
          projects: ['1'],
        },
      };

      const input = {
        project: 'some_project',
        notificationType: 'slack',
        notificationName: 'some_slack',
      };

      const ret = Sql.deleteProjectNotification(cred, input);
      expect(ret).toMatchSnapshot();
    });
  });
});
