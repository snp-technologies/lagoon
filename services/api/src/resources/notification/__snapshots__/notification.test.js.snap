// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sql deleteProjectNotification if not admin, should insert IN clause for project 1`] = `"delete from \`project_notification\` as \`pn\` where \`pn\`.\`pid\` in ('1') and \`p\`.\`name\` = 'some_project' and \`nt\`.\`name\` = 'some_slack'"`;

exports[`Sql selectAssignedProjectNotificationByName should create proper statement 1`] = `"select \`nt\`.*, \`pn\`.*, 'slack' as type from \`project_notification\` as \`pn\` JOIN notification_slack AS nt ON pn.nid = nt.id AND pn.type = 'slack' where \`nt\`.\`name\` = 'slack_foo'"`;

exports[`Sql selectCustomer should create a proper query 1`] = `"select * from \`notification_slack\` where \`name\` = 'n1'"`;

exports[`Sql selectNotificationsByTypeByProjectId if admin, should create statement for notification_slack table without IN clause 1`] = `"select \`nt\`.*, \`pn\`.\`type\` from \`project_notification\` as \`pn\` JOIN notification_slack AS nt ON pn.nid = nt.id AND pn.type = 'slack' where \`pn\`.\`pid\` = 3"`;

exports[`Sql selectNotificationsByTypeByProjectId if user, should create statement for notification_slack table with IN clause 1`] = `"select \`nt\`.*, \`pn\`.\`type\` from \`project_notification\` as \`pn\` JOIN notification_slack AS nt ON pn.nid = nt.id AND pn.type = 'slack' where \`pn\`.\`pid\` in ('3') and \`pn\`.\`pid\` = 3"`;

exports[`Sql selectProjectNotificationsWithoutAccess should create a proper statement 1`] = `"select \`pn\`.* from \`project_notification\` as \`pn\` inner join \`project\` as \`p\` on \`pn\`.\`pid\` = \`p\`.\`id\` where \`pn\`.\`nid\` in ('1') and \`pn\`.\`pid\` not in ('3', '4')"`;

exports[`Sql selectUnassignedNotificationsByType should create a proper query 1`] = `"select \`nt\`.*, 'rocketchat' as type from \`notification_rocketchat\` as \`nt\` left join project_notification AS pn ON pn.nid = nt.id AND pn.type = 'rocketchat' where pn.nid IS NULL and pn.pid IS NULL"`;

exports[`Sql updateNotificationSlack should create a proper query 1`] = `"update \`notification_slack\` set \`name\` = 'test', \`webhook\` = 'new-webhook' where \`name\` = 'n1'"`;
